src: upstream_patches_ui: Add bookmark feature

davidbtadokoro opened the pull request:
This commit adds a barebones implementation of the bookmark feature. This feature refers to the ability to choose a certain patch series from a public mailing list and save it to later consultation. After bookmarking a patch series, it will become available in the 'Bookmarked Patches' screen. At the moment, the only action available is to 'Unbookmark' a patch series, but it should be kind of simple to plug in new actions.

From the implementation point of view, kw needs to maintain a local database of all bookmarked patch series and this is done by creating and managing a file inside the kw data directory (much like kw kernel-config-manager). Although efficient and simple to manage, if the feature needs scaling (change/add entities/fields), a real DBMS should be considered. Also worth mentioning that, under the hood, bookmarking a patch series downloads it first in the same manner of the 'Download' action, i.e., saves the .mbox files inside a directory defined by the user in lore.config. Unbookmarking patches series also removes them first from local storage.

Closes: #796

Commit:
src: upstream_patches_ui: Add bookmark feature

This commit adds a barebones implementation of the bookmark feature.
This feature refers to the ability to choose a certain patch series from
a public mailing list and save it to later consultation. After
bookmarking a patch series, it will become available in the 'Bookmarked
Patches' screen. At the moment, the only action available is to
'Unbookmark' a patch series, but it should be kind of simple to plug in
new actions.

From the implementation point of view, kw needs to maintain a local
database of all bookmarked patch series and this is done by creating and
managing a file inside the kw data directory (much like kw
kernel-config-manager). Although efficient and simple to manage, if the
feature needs scaling (change/add entities/fields), a real DBMS should
be considered. Also worth mentioning that, under the hood, bookmarking a
patch series downloads it first in the same manner of the 'Download'
action, i.e., saves the .mbox files inside a directory defined by the
user in lore.config. Unbookmarking patches series also removes them
first from local storage.

Closes: #796

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -18,6 +18,12 @@ declare -gr MAILING_LISTS_PAGE='lore_main_page.html'
+declare -gr LORE_BOOKMARKED_SERIES='lore_bookmarked_series'
+
+# Path to bookmarked series file
+declare -g BOOKMARKED_SERIES_PATH="${KW_DATA_DIR}/${LORE_BOOKMARKED_SERIES}"
```
Add a new folder named `upstream_patches_ui`, i.e, `"${KW_DATA_DIR}/upstream_patches_ui/${LORE_BOOKMARKED_SERIES}"`.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+  local flag="$3"
+  local patch_file_name
+  local file_pattern
+  local command
```
`command` is a bash built-in; I recommend you use something like `cmd` instead.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+
+  file_pattern=$(printf '%s/*%s' "${download_dir_path}" "${patch_file_name}")
+
+  ls "${file_pattern}" 1> /dev/null 2>&1
```
`ls` command behavior may vary across distros; replace it with the `find` command or another option.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+  ls "${file_pattern}" 1> /dev/null 2>&1
+
+  if [[ "$ret" == 0 ]]; then
+    command="rm ${file_pattern}"
```
Always add a check before trying to run the `rm` command; that way you can ensure that the bash variable does not have anything like `/` or `/home/something`.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+
+  ls "${file_pattern}" 1> /dev/null 2>&1
+
+  if [[ "$ret" == 0 ]]; then
```
`ret`? You did not initialize this variable. Did I miss something?

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+    command="rm ${file_pattern}"
+    cmd_manager "$flag" "$command"
+  else
+    return 2 # ENOENT
```
You forgot to document this function return.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+# already exists
+function create_lore_bookmarked_file()
+{
+  if [[ ! -f "${BOOKMARKED_SERIES_PATH}" ]]; then
+    touch "${BOOKMARKED_SERIES_PATH}"
+  fi
```
Maybe you can use something like this:

```bash
  [[ -f "${BOOKMARKED_SERIES_PATH}" ]] && return

  touch "${BOOKMARKED_SERIES_PATH}"
```

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+
+# This function adds a given series to a local bookmark database for
+# series managed by kw. To achieve uniqueness for each series the function
+# assigns an id to it. Also a timestamp for informational purposes is also
```
/an id to it. Also a timestamp/an ID and timestamp/

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+  patch_id=$(printf '%s' "${target_patch}" | sha256sum | cut -d ' ' -f1)
+  timestamp=$(date '+%Y/%m/%d %H:%M')
+
+  if [[ $(grep -c "^${patch_id}" "${BOOKMARKED_SERIES_PATH}") == 0 ]]; then
```
Use `--count` instead of `-c`.
Additionally, I would remove the `grep` command from the `if` condition, and I would just use `"$?" == 0`.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+# @series_index: The index in the local bookmark database
+#
+# TODO:
+# - Find an alternative way to identify a series, this one may not be the most
+#   reliable.
```
The SQLite work will be very helpful for this.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+  local series_index="$1"
+
+  if [[ ! -f "${BOOKMARKED_SERIES_PATH}" ]]; then
+    return 2 # ENOENT
```
Don't forget to document returns.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+    return 2 # ENOENT
+  fi
+
+  sed -i "${series_index}d" "${BOOKMARKED_SERIES_PATH}"
```
Use `--in-place` instead of `-i`.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +464,152 @@ function download_series()
+
+  while IFS='' read -r series; do
+    IFS="${SEPARATOR_CHAR}" read -r -a columns <<< "$series"
+    timestamp="${columns[1]}"
+    patch_title="${columns[6]}"
+    patch_author="${columns[2]}"
```
At some point, we need to replace this `${columns[1]}` in the entire code with something more meaningful. Using an index is really confusing... maybe we can have a set of functions for that?

davidbtadokoro replied:
You are right. Even more considering that even in this same file they don't follow a pattern...

By a set a functions you are talking about something like `get_timestamp`, `get_patch_title` and so on to encapsulate this explicit `${columns[i]}`, right?

davidbtadokoro replied:
For sure. The issue regarding the integration BD. Going to implement these functions.

rodrigosiqueira commented on tests/lib/lore_test.sh:
```diff
@@ -14,6 +14,31 @@ function oneTimeSetUp()
+function tearDown()
+{
+  # Empty file
+  : > "${BOOKMARKED_SERIES_PATH}"
```
You can delete this file in the `tearDown` instead of empty it; I believe this way is more readable.

davidbtadokoro replied:
Right! I chose this approach because an error was thrown when trying to delete it (cause the file didn't exist). Thinking now, it is most likely because of the command on the comment right below 🤣

rodrigosiqueira commented on tests/lib/lore_test.sh:
```diff
@@ -170,4 +195,102 @@ function test_extract_metadata_from_patch_title_rfc()
 
+function test_create_lore_bookmarked_file()
+{
+  rm "${BOOKMARKED_SERIES_PATH}"
```
Make sure `"${BOOKMARKED_SERIES_PATH}"` is the path that you want to remove before calling `rm`.

rodrigosiqueira commented on tests/lib/lore_test.sh:
```diff
@@ -170,4 +195,102 @@ function test_extract_metadata_from_patch_title_rfc()
+  local count
+
+  add_series_to_bookmark "${target_patch1}" "${download_dir_path}"
+  output=$(cat "${BOOKMARKED_SERIES_PATH}")
```
You can use something more optimal than `cat` here; check https://kworkflow.org/content/codingstyle.html#command-substitution-and-arithmetic-expression .

davidbtadokoro replied:
Nice!

rodrigosiqueira commented on tests/lib/lore_test.sh:
```diff
@@ -170,4 +195,102 @@ function test_extract_metadata_from_patch_title_rfc()
+
+function test_add_series_to_bookmark()
+{
+  local target_patch1="someseries1"
+  local target_patch2="someseries2"
+  local target_patch3="someseries3"
+  local download_dir_path="somedir"
```
Single-quotes.

rodrigosiqueira commented on tests/lib/lore_test.sh:
```diff
@@ -170,4 +195,102 @@ function test_extract_metadata_from_patch_title_rfc()
+
+  add_series_to_bookmark "${target_patch2}" "${download_dir_path}"
+  add_series_to_bookmark "${target_patch3}" "${download_dir_path}"
+  count=$(wc -l "${BOOKMARKED_SERIES_PATH}" | cut -d ' ' -f1)
```
/-l/--lines/ and /-d/--delimiter/

rodrigosiqueira reviewed (changes requested):
Hi,

Thanks a lot for this PR. I added some inline comments and follows some general comments:

* The UI has some hang if I try the bookmarked option without anything (e.g., the first time after installing kw). You need to verify if backport list is empty and just add a notification to the user about that and an option to return to the previous menu.
* After I bookmark a patch, it will have some loading period, and the bookmarked option will be empty again. After bookmarking, a patch kw must keep that option marked; otherwise, users may need clarification later by thinking they did not bookmark the specific path.

davidbtadokoro commented:
Hi @rodrigosiqueira ,
Sorry for the delay. Working on it right now

Commit:
src: upstream_patches_ui: Add message box screen

When kw tries to call list_patches, but there are no patches to be
displayed (i.e., the _target_array_list is empty), dialog just hangs.
Furthermore, there isn't any information displayed to the user regarding
the occurrence.

In this context, this commit adds a message box screen that can, not
only be used to fix this case above, but also be used for any exception
treatment or notification. For dialog, when this screen occurs, a custom
title and message and an OK button to return to the flow of execution
are displayed.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented:
This force push addresses the inline comments and the addition of a screen to handle the case when there were no bookmarked patches to be displayed.

Working on making the Bookmark check persistent.

Commit:
src: lib: dialog_ui: Add persistence to checklist options

When the user bookmarked a patch (series) the next screen didn't alter
the status of the chosen options. In this case, we want the 'Bookmark'
and 'Download' options already checked, to notify the user that the
bookmark worked (and also for futher visits to the patch).

In this context, this commits implements this behaviour by adding
another parameter to the create_simple_checklist function named
check_statuses that is an array reference with all the statuses of the
checks (if they are on/off). As such, this behaviour is not limited to
the 'Bookmark' and 'Download' actions on a patch (series).

Notes:
- As we plan to revise the 'Download' action, when we do it, we need to
  decouple the 'Bookmark' check status from the 'Download' one (right
  now, 'Download' only checks when 'Bookmark' is checked);
- We may want to reduce the number of parameters in the
  create_simple_checklist function (and others in dialog_ui.sh) as it
  reached 10 arguments with this commit.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented:
This last commit addresses the persistence of checks of a patch (series).

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -36,6 +45,13 @@ declare -gr SEPARATOR_CHAR='Æ'
 # it easy to undertand.
 declare -ag list_of_mailinglist_patches
 
+function create_lore_data_dir()
```
Add documentation.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -36,6 +45,13 @@ declare -gr SEPARATOR_CHAR='Æ'
+  [[ -d "${LORE_DATA_DIR}" ]] && return
+
+  mkdir -p "${LORE_DATA_DIR}"
+}
```
I was wondering about this function... what if we cannot create LORE_DATA_DIR? Maybe we should add some extra checks for `mkdir` and handle the rare case where we cannot create this folder.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +474,169 @@ function download_series()
+
+  create_lore_bookmarked_file
+
+  patch_id=$(printf '%s' "${target_patch}" | sha256sum | cut -d ' ' -f1)
```
Don't we need to add any specific package in the setup phase to ensure that sha256sum is available?

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -115,8 +113,73 @@ function dashboard_entry_menu()
+
+  action_list=('Unbookmark')
+
+  # The local bookmark database starting index is 1 and the index
+  # passed as argument starts at 0.
+  ((series_index++))
```
Just to be safe and based on your comment, I recommend you to check `series_index` before trying to increment it.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -204,7 +271,9 @@ function show_series_details()
         case "$option" in
           'Bookmark')
-            printf 'TODO' # TODO
+            create_loading_screen_notification "Bookmarking patch(es)"$'\n'"- ${patch_title}"
+            download_series "$total_patches" "$patch_url" "${lore_config['download_to']}" "$patch_title"
+            add_series_to_bookmark "${target_patch}" "${lore_config['download_to']}"
```
If any of these functions fails, you should check and provide a message to the user that lets them know that the operation failed for some reason.

rodrigosiqueira replied:
+1 again in this comment.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -113,14 +113,19 @@ function dashboard_entry_menu()
   screen_sequence['RETURNING']=''
 
-  list_patches 'Bookmarked patches' bookmarked_series "${screen_sequence['SHOW_SCREEN']}"
+  fallback_message='kw could not find any bookmarked patches.'$'\n'
```
Could you add another `\n` to add a blank line between this message and the next one?

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -113,14 +113,19 @@ function dashboard_entry_menu()
-  list_patches 'Bookmarked patches' bookmarked_series "${screen_sequence['SHOW_SCREEN']}"
+  fallback_message='kw could not find any bookmarked patches.'$'\n'
+  fallback_message+='Try bookmarking patches in the menu "Registered mailing list"'
+  list_patches 'Bookmarked patches' bookmarked_series "${screen_sequence['SHOW_SCREEN']}" \
+    "${fallback_message}"
```
Please make this a single-line command; it is more readable.

rodrigosiqueira reviewed (changes requested):
Very nice PR!
I added some inline comments.

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -458,3 +474,169 @@ function download_series()
+
+  if [[ "$ret" == 0 ]]; then
+    while IFS=' ' read -r file; do
+      # Just to avoid trying to remove fs root
```
Nice! Yes, this is a fairly common task in the project 😆

Commit:
tests: kwlib_test: Suppress wrong ShellCheck fail

When we declare nested local functions (e.g., when we want to mock a
function for testing) ShellCheck fails in the rule SC2137 (unreachable
command). In that context, this commit fix the error by suppressing the
rule in those cases.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: kwlib: Add function for safety checking paths

It is a common task to remove files and directories. To avoid dangerous
removals, we need to check if the path is safe to remove. This commit
adds a predicate function named is_safe_path_to_remove that checks a
path safety.

An alternate approach would be to also encapsulate the removal, but
the chosen one gives the caller more flexibility.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented on src/kwlib.sh:
```diff
@@ -680,3 +680,21 @@ function get_current_env_name()
+  local path="$1"
+
+  [[ ! -e "$path" ]] && return 2 # ENOENT
+  [[ "$path" == '/' ]] && return 1
```
Please, document 1 as EPERM.

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -458,3 +487,167 @@ function download_series()
+# already exists
+function create_lore_bookmarked_file()
+{
+  create_lore_data_dir
```
In this new implementation, the above function can fail. Please check if it succeeds before creating a new file.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -115,8 +113,78 @@ function dashboard_entry_menu()
+      for option in "${selected_options[@]}"; do
+        case "$option" in
+          'Unbookmark')
+            delete_series_from_local_storage "${download_dir_path}" "${patch_title}"
+            remove_series_from_bookmark_by_index "${series_index}"
```
Correct me if I'm wrong, but these functions can fail. If so, could you please check the return value? I think we need to notify users in case of an error.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -276,13 +282,19 @@ function show_series_details()
-            create_loading_screen_notification "Bookmarking patch(es)"$'\n'"- ${patch_title}"
-            download_series "$total_patches" "$patch_url" "${lore_config['download_to']}" "$patch_title"
-            add_series_to_bookmark "${target_patch}" "${lore_config['download_to']}"
+            create_loading_screen_notification 'Bookmarking patch(es)'$'\n'"- ${patch_title}"
+            {
+              download_series "$total_patches" "$patch_url" "${lore_config['download_to']}" "$patch_title" &&
+                add_series_to_bookmark "${target_patch}" "${lore_config['download_to']}"
+            } || {
+              create_message_box 'Error' 'Could not bookmark patch(es)'$'\n'"- ${patch_title}"
+            }
```
Please ignore my comments; now I can see that you are checking the return values. However, could you use a simple `if` condition here? It should be more readable and aligned with the code style.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -276,13 +282,19 @@ function show_series_details()
-            create_loading_screen_notification "Downloading patch(es)"$'\n'"- ${patch_title}"
-            download_series "$total_patches" "$patch_url" "${lore_config['download_to']}" "$patch_title"
+            create_loading_screen_notification 'Downloading patch(es)'$'\n'"- ${patch_title}"
+            download_series "$total_patches" "$patch_url" "${lore_config['download_to']}" "$patch_title" || {
+              create_message_box 'Error' 'Could not download patch(es)'$'\n'"- ${patch_title}"
+            }
```
Same here... can we use `if`?

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -305,15 +317,24 @@ function show_series_details()
 
-  create_menu_options "${menu_title}" '' '_target_array_list' 1
+  if [[ -z "${_target_array_list}" ]]; then
+    create_message_box "${menu_title}" "${fallback_message}"
+    ret="$?"
+    screen_sequence['SHOW_SCREEN']='dashboard'
+    return "$ret"
```
Move `secreen_sequence`  right after the `if` condition, and remove the `ret` variable part in favor of just returning `$?`.

davidbtadokoro replied:
Sorry, I didn't understand the move of `screen_sequence`. When `list_patches` doesn't have patches to list, after the notification screen, the next screen is Dashboard. Otherwise, the next screen will be determined in the switch-case below. Moving `screen_sequence` after the `if` wouldn't be redundant?

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -305,15 +317,24 @@ function show_series_details()
+    ret="$?"
+    screen_sequence['SHOW_SCREEN']='dashboard'
+    return "$ret"
+  else
+    create_menu_options "${menu_title}" '' '_target_array_list' 1
+  fi
```
Since you added a return in the `if` condition, you don't need this else. You can call `create_menu_option` directly.

rodrigosiqueira commented:
Very nice set of changes! Thanks a lot for all of these improvements.

Merged into the unstable branch.
